diff -Nru a/bin/carbon-aggregator.py b/bin/carbon-aggregator.py
--- a/bin/carbon-aggregator.py  2011-05-19 05:28:30.000000000 +0100
+++ b/bin/carbon-aggregator.py  2011-05-19 05:29:21.000000000 +0100
@@ -20,19 +20,17 @@
 # Figure out where we're installed
 BIN_DIR = dirname( os.path.abspath(__file__) )
 ROOT_DIR = dirname(BIN_DIR)
-STORAGE_DIR = join(ROOT_DIR, 'storage')
-LOG_DIR = join(STORAGE_DIR, 'log', 'carbon-aggregator')
-LIB_DIR = join(ROOT_DIR, 'lib')
-CONF_DIR = join(ROOT_DIR, 'conf')
+STORAGE_DIR = '/var/lib/carbon'
+LOG_DIR = '/var/log/carbon'
+CONF_DIR = '/etc/carbon'
 
-sys.path.insert(0, LIB_DIR)
 os.environ['GRAPHITE_ROOT'] = ROOT_DIR
 
 # Parse command line options
 parser = optparse.OptionParser(usage='%prog [options] <start|stop|status>')
 parser.add_option('--debug', action='store_true', help='Run in the foreground, log to stdout')
 parser.add_option('--profile', help='Record performance profile data to the given file')
-parser.add_option('--pidfile', default=join(STORAGE_DIR, '%s.pid' % program.split('.')[0]), help='Write pid to the given file')
+parser.add_option('--pidfile', default=join('/var/run', '%s.pid' % program.split('.')[0]), help='Write pid to the given file')
 parser.add_option('--config', default=join(CONF_DIR, 'carbon.conf'), help='Use the given config file')
 parser.add_option('--rules', default=join(CONF_DIR, 'aggregation-rules.conf'), help='Use the give aggregation rules file')
 parser.add_option('--logdir', default=LOG_DIR, help="Write logs in the given directory")
diff -Nru a/bin/carbon-cache.py b/bin/carbon-cache.py
--- a/bin/carbon-cache.py 2011-05-19 05:28:30.000000000 +0100
+++ b/bin/carbon-cache.py 2011-05-19 05:29:21.000000000 +0100
@@ -38,14 +38,11 @@
 # Figure out where we're installed
 BIN_DIR = dirname( os.path.abspath(__file__) )
 ROOT_DIR = dirname(BIN_DIR)
-STORAGE_DIR = join(ROOT_DIR, 'storage')
-LOG_DIR = join(STORAGE_DIR, 'log', 'carbon-cache')
-LIB_DIR = join(ROOT_DIR, 'lib')
-CONF_DIR = join(ROOT_DIR, 'conf')
+STORAGE_DIR = '/var/lib/carbon'
+LOG_DIR = '/var/log/carbon'
+CONF_DIR = '/etc/carbon'
 __builtins__.CONF_DIR = CONF_DIR # evil I know, but effective.
 
-sys.path.insert(0, LIB_DIR)
-
 # Capture useful debug info for this commonly reported problem
 try:
   import carbon
@@ -61,7 +58,7 @@
 parser = optparse.OptionParser(usage='%prog [options] <start|stop|status>')
 parser.add_option('--debug', action='store_true', help='Run in the foreground, log to stdout')
 parser.add_option('--profile', help='Record performance profile data to the given file')
-parser.add_option('--pidfile', default=join(STORAGE_DIR, '%s.pid' % program.split('.')[0]), help='Write pid to the given file')
+parser.add_option('--pidfile', default=join('/var/run', '%s.pid' % program.split('.')[0]), help='Write pid to the given file')
 parser.add_option('--config', default=join(CONF_DIR, 'carbon.conf'), help='Use the given config file')
 parser.add_option('--logdir', default=LOG_DIR, help="Write logs in the given directory")
 
diff -Nru a/bin/carbon-relay.py b/bin/carbon-relay.py
--- a/bin/carbon-relay.py 2011-05-19 05:28:30.000000000 +0100
+++ b/bin/carbon-relay.py 2011-05-19 05:29:21.000000000 +0100
@@ -32,22 +32,19 @@
 
 
 # Figure out where we're installed
-BIN_DIR = dirname(__file__)
+BIN_DIR = dirname( os.path.abspath(__file__) )
 ROOT_DIR = dirname(BIN_DIR)
-STORAGE_DIR = join(ROOT_DIR, 'storage')
-LOG_DIR = join(STORAGE_DIR, 'log', 'carbon-relay')
-LIB_DIR = join(ROOT_DIR, 'lib')
-CONF_DIR = join(ROOT_DIR, 'conf')
+STORAGE_DIR = '/var/lib/carbon'
+LOG_DIR = '/var/log/carbon'
+CONF_DIR = '/etc/carbon'
 __builtins__.CONF_DIR = CONF_DIR
 
-sys.path.insert(0, LIB_DIR)
-
 
 # Parse command line options
 parser = optparse.OptionParser(usage='%prog [options] <start|stop|status>')
 parser.add_option('--debug', action='store_true', help='Run in the foreground, log to stdout')
 parser.add_option('--profile', help='Record performance profile data to the given file')
-parser.add_option('--pidfile', default=join(STORAGE_DIR, '%s.pid' % program.split('.')[0]), help='Write pid to the given file')
+parser.add_option('--pidfile', default=join('/var/run', '%s.pid' % program.split('.')[0]), help='Write pid to the given file')
 parser.add_option('--config', default=join(CONF_DIR, 'carbon.conf'), help='Use the given config file')
 parser.add_option('--rules', default=join(CONF_DIR, 'relay-rules.conf'), help='Use the give relay rules file')
 parser.add_option('--logdir', default=LOG_DIR, help="Write logs in the given directory")
