diff -Nru a/conf/carbon.conf.example b/conf/carbon.conf.example
--- a/conf/carbon.conf.example  2011-05-19 05:28:30.000000000 +0100
+++ b/conf/carbon.conf.example  2011-05-19 05:28:48.000000000 +0100
@@ -1,10 +1,10 @@
 [cache]
-LOCAL_DATA_DIR = /opt/graphite/storage/whisper/
+LOCAL_DATA_DIR = /var/lib/carbon/whisper/
 
 # Specify the user to drop privileges to
 # If this is blank carbon runs as the user that invokes it
 # This user must have write access to the local data directory
-USER = 
+USER = carbon
 
 # Limit the size of the cache to avoid swapping or becoming CPU bound.
 # Sorts and serving cache queries gets more expensive as the cache grows.
