diff -ru a/webapp/graphite/settings.py b/webapp/graphite/settings.py
--- a/webapp/graphite/settings.py 2011-04-01 20:48:43.000000000 +0100
+++ b/webapp/graphite/settings.py 2011-05-14 23:48:34.000000000 +0100
@@ -22,16 +22,15 @@
 # Filesystem layout (all directores should end in a /)
 WEB_DIR = dirname( abspath(__file__) ) + '/'
 WEBAPP_DIR = dirname( dirname(WEB_DIR) ) + '/'
-GRAPHITE_ROOT = dirname( dirname(WEBAPP_DIR) ) + '/'
-CONF_DIR = GRAPHITE_ROOT + 'conf/'
-CONTENT_DIR = WEBAPP_DIR + 'content/'
-STORAGE_DIR = GRAPHITE_ROOT + 'storage/'
-WHISPER_DIR = STORAGE_DIR + 'whisper/'
-RRD_DIR = STORAGE_DIR + 'rrd/'
-LISTS_DIR = STORAGE_DIR + 'lists/'
+CONF_DIR = '/etc/graphite/'
+CONTENT_DIR = '/usr/share/graphite/webapp/content/'
+STORAGE_DIR = '/var/lib/graphite/'
+WHISPER_DIR = '/var/lib/carbon/whisper/'
+RRD_DIR = '/var/lib/carbon/rrd/'
+LISTS_DIR = '/var/lib/carbon/lists/'
 INDEX_FILE = STORAGE_DIR + 'index'
 WHITELIST_FILE = LISTS_DIR + 'whitelist'
-LOG_DIR = STORAGE_DIR + 'log/webapp/'
+LOG_DIR = '/var/log/graphite/'
 THIRDPARTY_DIR = WEB_DIR + 'thirdparty/'
 CLUSTER_SERVERS = []
 
