diff --git a/stats.js b/stats.js
index f76d173..f84cbc5 100644
--- a/stats.js
+++ b/stats.js
@@ -68,9 +68,8 @@ config.configFile(process.argv[2], function (config, oldConfig) {
       var key;
 
       for (key in counters) {
-        var value = counters[key] / (flushInterval / 1000);
+        var value = counters[key];
         var message = 'stats.' + key + ' ' + value + ' ' + ts + "\n";
-        message += 'stats_counts.' + key + ' ' + counters[key] + ' ' + ts + "\n";
         statString += message;
         counters[key] = 0;
 
