From c4a5b224da4cdf312268e9ccc8d278f3175ff77e Mon Sep 17 00:00:00 2001
From: Vicent Marti <tanoku@gmail.com>
Date: Wed, 17 Aug 2011 01:27:38 +0200
Subject: [PATCH] More lax checking of timezones

---
 fsck.c |   18 +++++++++++-------
 1 files changed, 11 insertions(+), 7 deletions(-)

diff --git a/fsck.c b/fsck.c
index 3d05d4a..e6af207 100644
--- a/fsck.c
+++ b/fsck.c
@@ -245,14 +245,18 @@ static int fsck_ident(char **ident, struct object *obj, fsck_error error_func)
 	if (**ident != ' ')
 		return error_func(obj, FSCK_ERROR, "invalid author/committer line - bad date");
 	(*ident)++;
-	if ((**ident != '+' && **ident != '-') ||
-	    !isdigit((*ident)[1]) ||
-	    !isdigit((*ident)[2]) ||
-	    !isdigit((*ident)[3]) ||
-	    !isdigit((*ident)[4]) ||
-	    ((*ident)[5] != '\n'))
+
+	if (**ident != '+' && **ident != '-')
 		return error_func(obj, FSCK_ERROR, "invalid author/committer line - bad time zone");
-	(*ident) += 6;
+	(*ident)++;
+
+	do {
+		if (!isdigit(**ident))
+			return error_func(obj, FSCK_ERROR, "invalid author/committer line - bad time zone");
+		(*ident)++;
+	} while (**ident != '\n');
+	(*ident)++;
+
 	return 0;
 }
 
-- 
1.7.5.4

