--- a/sshd.c
+++ b/sshd.c
@@ -259,6 +259,9 @@
 /* Unprivileged user */
 struct passwd *privsep_pw = NULL;
 
+/* Timestamp string when client request started */
+char *accept_timestamp;
+
 /* Prototypes for various functions defined later in this file. */
 void destroy_sensitive_data(void);
 void demote_sensitive_data(void);
@@ -1806,6 +1916,15 @@
 	/* This is the child processing a new connection. */
 	setproctitle("%s", "[accepted]");

+	{
+		struct timespec tv;
+		clock_gettime(CLOCK_MONOTONIC, &tv);
+		accept_timestamp = xmalloc(32);
+		snprintf(accept_timestamp, 32,
+			"%ld%ld", tv.tv_sec, tv.tv_nsec/1000000);
+	}
+
+
 	/*
 	 * Create a new session and process group since the 4.4BSD
 	 * setlogin() affects the entire process group.  We don't
--- a/session.c
+++ b/session.c
@@ -134,6 +134,7 @@
 extern void destroy_sensitive_data(void);
 extern Buffer loginmsg;
 extern char *last_fingerprint;
+extern char *accept_timestamp;
 
 /* original command from peer. */
 const char *original_command = NULL;
@@ -1235,6 +1236,8 @@
 	if (original_command)
 		child_set_env(&env, &envsize, "SSH_ORIGINAL_COMMAND",
 		    original_command);
+	child_set_env(&env, &envsize, "GIT_SOCKSTAT_VAR_SSHD_ACCEPT_TIME",
+		accept_timestamp);
 
 #ifdef _UNICOS
 	if (cray_tmpdir[0] != '\0')
--- a/Makefile.in
+++ b/Makefile.in
@@ -48,7 +48,7 @@
 CFLAGS=@CFLAGS@
 CPPFLAGS=-I. -I$(srcdir) @CPPFLAGS@ $(PATHS) @DEFS@
 LIBS=@LIBS@
-SSHDLIBS=@SSHDLIBS@
+SSHDLIBS=@SSHDLIBS@ -lrt
 LIBEDIT=@LIBEDIT@
 AR=@AR@
 AWK=@AWK@
