diff --git a/sshd.c b/sshd.c
index 9e03cf3..b51c642 100644
--- a/sshd.c
+++ b/sshd.c
@@ -1209,6 +1209,13 @@ server_accept_loop(int *sock_in, int *sock_out, int *newsock, int *config_s)
 				 * We break out of the loop to handle
 				 * the connection.
 				 */
+				struct timespec tv;
+				char t[32];
+				clock_gettime(CLOCK_MONOTONIC, &tv);
+				snprintf(t, 32, "%ld%ld", tv.tv_sec, tv.tv_nsec/1000000);
+				t[31] = '\0';
+				setenv("GIT_SOCKSTAT_VAR_SSHD_ACCEPT_TIME", t, 1);
+
 				platform_post_fork_child();
 				startup_pipe = startup_p[1];
 				close_startup_pipes();
--- a/Makefile.in
+++ b/Makefile.in
@@ -48,7 +48,7 @@
 CFLAGS=@CFLAGS@
 CPPFLAGS=-I. -I$(srcdir) @CPPFLAGS@ $(PATHS) @DEFS@
 LIBS=@LIBS@
-SSHDLIBS=@SSHDLIBS@
+SSHDLIBS=@SSHDLIBS@ -lrt
 LIBEDIT=@LIBEDIT@
 AR=@AR@
 AWK=@AWK@
