diff --git a/carbon/setup.cfg b/carbon/setup.cfg
deleted file mode 100644
index aeccfb2..0000000
--- a/carbon/setup.cfg
+++ /dev/null
@@ -1,3 +0,0 @@
-[install]
-prefix = /opt/graphite
-install-lib = %(prefix)s/lib
diff --git a/carbon/setup.py b/carbon/setup.py
index d9a445f..9c9c975 100644
--- a/carbon/setup.py
+++ b/carbon/setup.py
@@ -26,9 +27,8 @@ setup(
   description='Backend data caching and persistence daemon for Graphite',
   packages=['carbon', 'carbon.aggregator'],
   package_dir={'' : 'lib'},
-  scripts=glob('bin/*'),
+  scripts=glob('bin/*.py'),
   package_data={ 'carbon' : ['*.xml'] },
-  data_files=storage_dirs + conf_files,
   install_requires=['twisted', 'txamqp'],
   **setup_kwargs
 )

